src/borrow_tracker/tree_borrows/mod.rs

```diff
@@ -317,32 +317,33 @@ trait EvalContextPrivExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let mut tree_borrows = alloc_extra.borrow_tracker_tb().borrow_mut();
         let prot = new_perm.protector.is_some();
 
-        // Store initialized permissions and their corresponding range.
+        // Store initial permissions and their corresponding range.
         let mut perms_map: RangeMap<LocationState> = RangeMap::new(
-            base_offset + ptr_size,
+            ptr_size,
             LocationState::new_uninit(
                 Permission::new_disabled(),
                 foreign_access_skipping::IdempotentForeignAccess::None,
             ),
         );
+        // Keep track of whether the node has any part that allows for interior mutability.
+        let mut is_frozen = true;
 
-        let NewPermission { freeze_perm, freeze_access, nonfreeze_perm, nonfreeze_access, .. } =
-            new_perm;
         this.visit_freeze_sensitive(place, ptr_size, |mut range, frozen| {
+            is_frozen = frozen;
```

Comment:
```suggestion
            is_frozen = is_frozen || frozen;
```
Otherwise this does not seem to match the comment?

Reply:
Ah right. We'd need a bunch of negations. I suggest you rename the variable to `has_unsafe_cell`, then the logic becomes a lot simpler -- `has_unsafe_cell = has_unsafe_cell || !frozen`.